Even faster .NET Parser

philipmat opened the pull request:
| File | Record Count | Python | C# v1 | C# v2 |
| --- | ---: | :---: | :---: | :---: |
| discogs_20200806_artists.xml.gz | 7,046,615 | 6:22| 2:35 | 0:28 - 13x / 5x |
| discogs_20200806_labels.xml.gz | 1,571,873 | 1:15 | 0:22 | 0:05 - 15x / 4x |
| discogs_20200806_masters.xml.gz | 1,734,371 | 3:56 | 1:57 | 0:31 - 7x / 4x |
| discogs_20200806_releases.xml.gz | 12,867,980 | 1:45:16 | 42:38 | 15:17 - 7x / 3x |

That's just a part of the story.
Parallel, processing is the other. 
<img width="1032" alt="Screen Shot 2020-09-13 at 11 36 31 AM" src="https://user-images.githubusercontent.com/709928/93023606-bd2dc300-f5b5-11ea-8ab6-d273cbfffe76.png">

As the screenshot above shows, the .NET version allows processing multiple files in parallel, achieving a 20x speedup from the Python version in this case.

philipmat committed:
uses direct identification of nodes in reader

philipmat committed:
renames exporter interface and method to remove csv

philipmat committed:
makes individual classes responsible for populating themselves

philipmat committed:
parses artists (TODO: members)

philipmat committed:
avoids creating satellite assemblies

philipmat committed:
parses members as well

philipmat committed:
excludes some temp python folders

philipmat committed:
reverts changes by an overly eager vscode helper

philipmat committed:
parses groups

in part it's done to have a more accurate parser

philipmat committed:
parses sublabels

philipmat committed:
skips nodes we don't (yet) parse/export

philipmat committed:
tests artist being populated from reader and sample file

philipmat committed:
allows for both deserialize and parse to operate on the same file

philipmat committed:
removes unnecessary usings

philipmat committed:
sublabel populate tests

philipmat committed:
asserts on images

philipmat committed:
reads sublabels

philipmat committed:
reads groups

philipmat committed:
reduces the number of images on label

philipmat committed:
reads master, except for artist and videos

philipmat committed:
parses master artists

philipmat committed:
reads videos and hardens artist reading

philipmat committed:
adds two artists for test

philipmat committed:
implements IsValid

philipmat committed:
parser master artists and videos correctly

philipmat committed:
Merge branch 'develop' into dotnet_parser_specific_xml_parsing

# Conflicts:
#	alternatives/dotnet/discogs/Program.cs

philipmat committed:
updates RecordCounter with develop changes

philipmat committed:
Merge branch 'develop' into dotnet_parser_specific_xml_parsing

philipmat committed:
images provide a factory method

philipmat committed:
reads top level artist tags and images

philipmat committed:
ignores whitespace on reading

philipmat committed:
parses videos

philipmat committed:
parses genres and styles

philipmat committed:
parses identifiers

philipmat committed:
exits at the end of switch on release node

philipmat committed:
parses labels

philipmat committed:
prepares asserts for future tags

philipmat committed:
reads formats

philipmat committed:
handles format having no descriptions

philipmat committed:
parses artists and extra artists

philipmat committed:
parses companies

philipmat committed:
parser subtracks

philipmat committed:
renames test to indicare it's release-specific

philipmat committed:
outputs track sequence in track artists

philipmat committed:
release-style parsing for artists

philipmat committed:
release-style parsing for masters

philipmat committed:
skips end type element for faster parsing

philipmat committed:
parses labels, release style

philipmat committed:
renames variable to more generic name

philipmat committed:
makes xml reader settings public to use in tests

philipmat committed:
processes multiple files in parallel

philipmat committed:
allows running the xml deserializer version with --v1

philipmat committed:
adds a marker for troubleshooting tests

philipmat committed:
troubleshooting tests

MuleaneEve commented:
Very cool to go from a process that used to take almost 2 hours to barely 15 minutes :)

philipmat committed:
switches FluentAssertions for AwesomeAssertions and upgrades all project references

philipmat committed:
config files for .NET coding guidelines and AI agents

philipmat committed:
adds an editor config file for consistent styles

philipmat committed:
small improvements to README

philipmat committed:
adds various words to team-shared dictionary

philipmat committed:
post-merge fixes

philipmat committed:
enables implicit usings

philipmat committed:
removes unused usings

philipmat committed:
file-scoped namespaces

philipmat committed:
global usings

philipmat committed:
implements modern syntax and addresses analyzer recommendations

philipmat committed:
todo on bringing in a parser library

philipmat committed:
renames class to reflect it uses an XmlSerializer

philipmat committed:
puts interface in own file

philipmat committed:
documentation for exportable class

philipmat committed:
minor formatting

philipmat committed:
updates counts to more recent numbers so the progress bar can work better

philipmat committed:
small formatting

philipmat committed:
starts aligning names with C# standards

philipmat committed:
fixes xUnit3003 warning

philipmat committed:
uses C# standard naming conventions for classes and properties

philipmat committed:
top level namespace for Artist

philipmat committed:
C# naming conventions for labels

philipmat committed:
C# conventions for labels

philipmat committed:
C# conventions for Master

philipmat committed:
converts Master to C# standards

philipmat committed:
updates GH actions to latest versions

philipmat committed:
Merge branch 'develop' into dotnet_parser_specific_xml_parsing

philipmat committed:
uses a case-insensitive dictionary for statistics

philipmat committed:
counts records in csv files

philipmat committed:
Merge branch 'develop' into dotnet_parser_specific_xml_parsing

# Conflicts:
#	README.md

philipmat committed:
notes on implementation and lack of coherence in the sample files

philipmat committed:
renames file to match class name

philipmat committed:
test files for a C# test

the csv samples are the same as the python version uses, and the csv files are the ones produce by the `test_extract.py`

philipmat committed:
tests for comparing exported csv files to baseline

philipmat committed:
adds missing fields to export

export now has parity to python version

philipmat committed:
passes in file name to make tests easier to read

philipmat committed:
missed a quote

philipmat committed:
allows the discog files in the samples folder

philipmat committed:
adds the samples files used for testing

philipmat committed:
outputs temporary directory path (to be able to stop and copy the files)
